### PR TITLE
[FLINK-33470] Implement restore tests for Join node

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for Join node
* Adds a clearData method to clean up the TestValuesFactory between RestoreTests.

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for Join node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by jnh5y** [comment]
This PR adds back the JoinRestore tests (https://github.com/apache/flink/pull/23680) which were reverted (https://github.com/apache/flink/pull/23861) after CI failures.  

The issues was that other tests using the same factory name in the TestValuesFactory were reused between tests and the RestoreTestBase was not cleaning them up.  

This PR adds a method to clean up the TestValuesFactory between tests.  This should be sufficient as long as multiple RestoreTests are not run concurrently.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java [hunk 1]
```diff
@@ -117,6 +122,11 @@ public EnumSet<TestKind> supportedRunSteps() {
         return EnumSet.of(TestKind.SQL);
     }
 
+    @BeforeEach
+    public void clearData() {
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java:126]: Nice find!

I think we should be unified with how it's done at other places. `JsonPlanTestBase` does it in `AfterEach`. If we reuse a name between those two, `JsonPlanTestBase` can fail if executed after `RestoreTestBase`.

Could we also clear the `localRawResultsObservers` here?

**Review by dawidwys — CHANGES_REQUESTED** [review]

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+                    .producedBeforeRestore(
+                            Row.of(1, 1L, "BakerBaker"),
+                            Row.of(2, 2L, "HeHe"),
+                            Row.of(3, 2L, "HeHe"))
+                    .producedAfterRestore(Row.of(2, 1L, "PostRestoreRight"))
+                    .build();
+
+    static {
+        NON_WINDOW_INNER_JOIN =
+                TableTestProgram.of("non-window-inner-join", "test non-window inner join")
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java:110]: I missed it previously. Could you please adapt the id according to:
https://github.com/apache/flink/blob/52d8d3583e5c989da84126a8805ab335408c46c2/flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/test/program/TableTestProgram.java#L117

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java (hunk 1 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java:126]: A food for thought. Would it make sense to add some unique identifier to all the sinks/sources in `RestoreTestBase`? We could use e.g. the test id or generate some UUID for the duration of a test run.

Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/RestoreTestBase.java:126]: >A food for thought. Would it make sense to add some unique identifier to all the sinks/sources in RestoreTestBase? We could use e.g. the test id or generate some UUID for the duration of a test run.

I thought about trying to do that.  Seems like we'd have to do it manually in the test cases or update the SQL somehow if we were doing it in the code.  I didn't see an easy way to sort that out.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java [hunk 3]
```diff
@@ -107,7 +107,7 @@ public class JoinTestPrograms {
 
     static {
         NON_WINDOW_INNER_JOIN =
-                TableTestProgram.of("non-window-inner-join", "test non-window inner join")
+                TableTestProgram.of("join-non-window-inner-join", "test non-window inner join")
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java:110]: Hope you won't find it nitpicking, but can we leave just a single occurence of `join`? I think `join-non-window-inner` still makes sense, but I am not a native speaker. If you find the duplication better leave it as is.

Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java:110]: Yeah, I don't like the duplication either...  Both options are unfortunate; I find this one slightly better.  Can we keep it as-is?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java (hunk 2 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java]: Yes!

**Review by dawidwys — COMMENTED** [review]
let's wait for a green CI

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java (hunk 3 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java:110]: 👍

**Pull request closed without merging**